source/common/config/grpc_subscription_impl.h

```diff
@@ -13,35 +14,42 @@ namespace Config {
 
 template <class ResourceType>
 class GrpcSubscriptionImpl : public Config::Subscription<ResourceType>,
-                             Grpc::AsyncClientCallbacks<envoy::api::v2::DiscoveryResponse> {
+                             Grpc::AsyncClientCallbacks<envoy::api::v2::DiscoveryResponse>,
+                             Logger::Loggable<Logger::Id::config> {
 public:
   GrpcSubscriptionImpl(const envoy::api::v2::Node& node, Upstream::ClusterManager& cm,
                        const std::string& remote_cluster_name, Event::Dispatcher& dispatcher,
-                       const google::protobuf::MethodDescriptor& service_method)
+                       const google::protobuf::MethodDescriptor& service_method,
+                       SubscriptionStats stats)
       : GrpcSubscriptionImpl(
             node, std::unique_ptr<Grpc::AsyncClientImpl<envoy::api::v2::DiscoveryRequest,
                                                         envoy::api::v2::DiscoveryResponse>>(
                       new Grpc::AsyncClientImpl<envoy::api::v2::DiscoveryRequest,
                                                 envoy::api::v2::DiscoveryResponse>(
                           cm, remote_cluster_name)),
-            dispatcher, service_method) {}
+            dispatcher, service_method, stats) {}
 
   GrpcSubscriptionImpl(
       const envoy::api::v2::Node& node,
       std::unique_ptr<Grpc::AsyncClient<envoy::api::v2::DiscoveryRequest,
                                         envoy::api::v2::DiscoveryResponse>> async_client,
-      Event::Dispatcher& dispatcher, const google::protobuf::MethodDescriptor& service_method)
+      Event::Dispatcher& dispatcher, const google::protobuf::MethodDescriptor& service_method,
+      SubscriptionStats stats)
       : async_client_(std::move(async_client)), service_method_(service_method),
-        retry_timer_(dispatcher.createTimer([this]() -> void { establishNewStream(); })) {
+        retry_timer_(dispatcher.createTimer([this]() -> void { establishNewStream(); })),
+        stats_(stats) {
     request_.mutable_node()->CopyFrom(node);
   }
 
   void setRetryTimer() { retry_timer_->enableTimer(std::chrono::milliseconds(RETRY_DELAY_MS)); }
 
   void establishNewStream() {
+    ENVOY_LOG(debug, "Establishing new gRPC bidi stream for {}", service_method_.DebugString());
+    stats_.update_attempt_.inc();
     stream_ = async_client_->start(service_method_, *this, Optional<std::chrono::milliseconds>());
     if (stream_ == nullptr) {
-      // TODO(htuch): Track stats and log failures.
+      ENVOY_LOG(warn, "Unable to establish new stream");
+      stats_.update_failure_.inc();
```

Comment:
There is duplication here in terms of inc stat, calling onCofnigUpdateFailed, setting retry timer, etc. similar to below. Can we merge into a common local failure helper?